Add post author selector enhanced with ComboboxControl

adamsilverstein opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Add accessible search with debounced api requests for the post author dropdown.

Built on the ComboBox being added in https://github.com/WordPress/gutenberg/pull/19657 which relies on Downshift.

Supersedes https://github.com/WordPress/gutenberg/pull/7385.

## How has this been tested?
* Tested with a dozen users, by adding them with wp-cli: `wp user generate --role=editor --count=12`

Regular dropdown shows.

* Tested with 10100 users by adding them with wp-cli: `wp user generate --role=editor --count=10100`

Enhanced selector shows. Typing will begin searching in the selector.

Note: to clear out generated users, I used `wp user delete $(wp user list --role=editor --field=ID) --reassign=2`


## Screenshots <!-- if applicable -->
![Screen Recording 2020-06-19 at 11 27 AM](https://user-images.githubusercontent.com/2676022/85163422-06bd7a80-b220-11ea-96cd-88908e70ce55.gif)

## Types of changes
* Only load the first 100 users from the API at load.
* Load the post author object by id at load.
* Use a debounced search to query the API for matching users while the user types.
* Question: should we start searching once the user types 2 characters?
* Use progressive enhancement to add the enhanced selector when there are more than 50 users to choose from.
* I haven't added any styles, I am assuming the underlying component will include these.
* I added a loading results state to show a spinner while fetching authors, should this be tracked in the store vs the component?

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

epiqueras committed:
Components: Implement a combobox control.

epiqueras committed:
Components: Update Downshift.

epiqueras committed:
Components: Update Downshift.

adamsilverstein committed:
New post author selector, initial scaffolding.

adamsilverstein committed:
New post author selector, async work.

adamsilverstein committed:
Complete initial implementation.

adamsilverstein committed:
Cleanup.

adamsilverstein committed:
Clean up user selection and search.

epiqueras commented on packages/components/src/combobox-control/index.js:
```diff
@@ -94,31 +106,37 @@ export default function ComboboxControl( {
-							) }
-							{ item.name }
-						</li>
+					( isLoading ? (
+						<span>
+							<Spinner /> { __( 'Loading' ) }
+						</span>
+					) : (
+						items.map( ( item, index ) => (
+							// eslint-disable-next-line react/jsx-key
+							<li
+								{ ...getItemProps( {
+									item,
+									index,
+									key: item.key,
+									className: classnames(
+										'components-combobox-control__item',
+										{
+											'is-highlighted':
+												index === highlightedIndex,
+										}
+									),
+									style: item.style,
+								} ) }
+							>
+								{ item === selectedItem && (
+									<Dashicon
+										icon="saved"
+										className="components-combobox-control__item-icon"
+									/>
+								) }
+								{ item.name }
+							</li>
+						) )
```
Can we show an overlay over the current items instead? The current interaction is very jarring.

adamsilverstein replied:
do you mean an overlay for the spinner?

epiqueras replied:
An overlay over the list of items instead of the spinner.

adamsilverstein replied:
ah, ok - so the list shows as faded/inactive while the new results are being fetched?

epiqueras replied:
Yes

adamsilverstein replied:
👍 will do

adamsilverstein replied:
I added this: the list grays out while the load happens. I noticed the resolved returned an empty list during resolution, so I added some logic in the parent to keep the list unchanged until it resolves (so the previously showing list is greyed out until the fetch completes. Please take a look.

adamsilverstein replied:
removed this bit for now.

epiqueras commented on packages/core-data/src/selectors.js:
```diff
@@ -75,6 +75,17 @@ export const getUserQueryResults = createSelector(
+ * @return {Array} Authors list.
+ */
+export function getAuthor( state ) {
+	return getUserQueryResults( state, 'author' );
```
It might make more sense to use `state.users.byId[ id ]` here.

adamsilverstein replied:
👍

adamsilverstein replied:
This only works if the post author is among the authors in the initial selection. We need to request the post author explicitly.

epiqueras replied:
What do you mean? The resolver would still run.

adamsilverstein replied:
the authors resolver gets up to 100 authors. the current post author may not be among them, so  pretty sure in that case `state.users.byId[ id ]` will not be defined?

epiqueras replied:
The single author resolver also populates `byId`.

adamsilverstein replied:
ah, ok makes sense. I will change this and verify it works as expected.

adamsilverstein replied:
updated

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Internal dependencies
+ */
+import ComboboxControl from '../../../../components/build/combobox-control/';
```
This can be imported from the components package.

adamsilverstein replied:
I tried and it didn't work, anything I need to do to have it recognized correctly? Do you need to merge the other PR first and publish the package?

adamsilverstein replied:
Any tips on how? When I try:

`import { ComboboxControl } from '@wordpress/components';`

I get an error at run time:
`React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.`

epiqueras replied:
The component is not exported yet. You can add it to the index of the components package.

adamsilverstein replied:
ok, i'll try to hunt that down - do you need to add that to your PR before merging?

epiqueras replied:
It was already merged.

adamsilverstein replied:
excellent, I saw that! I'll update the import and try again, should `import { ComboboxControl } from '@wordpress/components';`` work directly now or is there something else I need to do?

adamsilverstein replied:
got this working once i exported the component.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+import ComboboxControl from '../../../../components/build/combobox-control/';
+
+function EnhancedPostAuthor( { postAuthor, id, authors, onUpdateAuthor } ) {
+	const currentPostAuthor = postAuthor[ 0 ];
```
Why is the author an array?

adamsilverstein replied:
Because getAuthor( id ) currently returns an array, I can change this so the data resolves as a single entry.

adamsilverstein replied:
Changed the selector to return a single value.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+
+function EnhancedPostAuthor( { postAuthor, id, authors, onUpdateAuthor } ) {
+	const currentPostAuthor = postAuthor[ 0 ];
+	let initialAuthors = [];
+	authors.forEach( ( author, i ) => {
+		initialAuthors = [
+			...initialAuthors,
+			{
+				key: i,
+				name: author.name,
+				value: author.name,
+				id: author.id,
+			},
+		];
+	} );
```
This should be memoized and use `.map`, not `.forEach`.

adamsilverstein replied:
Ok - why may vs. forEach? I had map originally but we aren't returning anything so why the preference for map?

adamsilverstein replied:
switched to map, that is cleaner. Somehow didn't work originally.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+
+	// Ensure the current author is included in the initial dropdown list.
+	let foundAuthor = initialAuthors.findIndex(
+		( author ) => currentPostAuthor && currentPostAuthor.id === author.id
```
```suggestion
		( author ) => currentPostAuthor?.id === author.id
```

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+		( author ) => currentPostAuthor && currentPostAuthor.id === author.id
+	);
+
+	if ( currentPostAuthor && foundAuthor < 0 ) {
```
What happens when there is no `currentPostAuthor`?

adamsilverstein replied:
There is always a post author, for new posts this is the current user.

epiqueras replied:
Then why do you check if it's truthy in the conditional?

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+	);
+
+	if ( currentPostAuthor && foundAuthor < 0 ) {
+		currentPostAuthor.key = initialAuthors.length;
```
You can't mutate props.

adamsilverstein replied:
no longer present

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+
+	if ( currentPostAuthor && foundAuthor < 0 ) {
+		currentPostAuthor.key = initialAuthors.length;
+		initialAuthors = [ currentPostAuthor, ...initialAuthors ];
```
`unshift` would be more performant.

adamsilverstein replied:
updated, thanks.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	// The search result loading state.
+	const [ isLoadingSearchResults, setIsLoadingSearchResult ] = useState(
```
```suggestion
	const [ isLoadingSearchResults, setIsLoadingSearchResults ] = useState(
```

adamsilverstein replied:
no longer present

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+	);
+
+	// The currently selected author.
+	const [ selectedAuthor, setSelectedAuthor ] = useState( currentPostAuthor );
```
There's no need to mirror the controlled props in local state like this, `currentPostAuthor` updates.

adamsilverstein replied:
this is used to track the current selection so we know not to search when the typed (input) value matches the current user, may no longer be needed - i'll test removing.

adamsilverstein replied:
without this, once you select an author the search becomes active and the dropdown shows only your user. with this, the dropdown goes back to showing the default users. I will remove for now to simplify and we can consider the best user interaction.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+	const searchAuthors = debounce( ( query ) => {
+		const payload = '?search=' + encodeURIComponent( query );
+		setIsLoadingSearchResult( true );
+		apiFetch( { path: '/wp/v2/users' + payload } ).then( ( results ) => {
```
Doesn't this need `who=authors`?

epiqueras replied:
This also should be done in `useSelect`, `getAuthors` is built in a way that supports adding queries.

adamsilverstein replied:
I'll give using getAuthors a try here, I didn't know we could pass a query. Not sure what `who=authors` does I will review.

adamsilverstein replied:
Not seeing how `getAuthors` support adding queries? I will add that.

adamsilverstein replied:
I added a new searchAuthors selector and resolver, if we use the original getAuthors, it changes the authors property which we don't want.

I switched to `useSelect` which simplifies the code here. I'm no longer tracking loading state because of this change. Should I add something to the resolver to yield a 'SEARCHING' state until it completes... or any suggestions for how to track a loading state while the resolver is fetching the data?

epiqueras replied:
We don't need a `searchAuthors`, `getAuthors` needs a new query parameter that works like `getEntityRecords`'s.

adamsilverstein replied:
I tried this approach and can re-visit; with the current setup, I found getAuthors with parameter would overwrite original state authors, causing a prop change.

How about tracking the loading state question?

epiqueras replied:
`receiveUserQuery`'s first parameter is a query key.

See how entity record queries use packages/core-data/src/queried-data/get-query-parts.js to generate keys from the query.

>How about tracking the loading state question?

You should be able to use `select( 'core/data' ).isResolving`.

adamsilverstein replied:
> See how entity record queries use packages/core-data/src/queried-data/get-query-parts.js to generate keys from the query.

Excellent, thanks for the pointer.

> You should be able to use select( 'core/data' ).isResolving.

Brilliant, I'll try that.

adamsilverstein replied:
I still found I needed to construct a unique key, which is maybe fine - this accounts for an empty query.

adamsilverstein replied:
Added.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+			setAvailableAuthors(
+				results.map( ( author, i ) => ( {
+					key: i,
+					name: author.name,
+					value: author.name,
```
Why do we need to duplicate the name like this?

adamsilverstein replied:
I will double check.

adamsilverstein replied:
remove `value`

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+					key: i,
+					name: author.name,
+					value: author.name,
+					id: author.id,
```
Shouldn't the key be the id, and that way, this isn't needed, and render reconciliation does not rely on ordering?

adamsilverstein replied:
good suggestion, will change.

adamsilverstein replied:
updated.

epiqueras commented on packages/editor/src/components/post-author/enhanced-post-author.js:
```diff
@@ -0,0 +1,115 @@
+
+	return (
+		<ComboboxControl
+			id={ id }
```
`ComboboxControl` doesn't take an ID.

adamsilverstein replied:
If we use multiple `ComboboxControl` on a page, do we need to pass an id to ensure they have unique keys?

adamsilverstein replied:
removed for now

epiqueras replied:
>do we need to pass an id to ensure they have unique keys?

No

epiqueras commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -26,27 +26,48 @@ export class PostAuthor extends Component {
+						id={ selectId }
+						postAuthor={ postAuthor }
+						authors={ authors }
+						instanceId={ instanceId }
```
Why is this needed?

adamsilverstein replied:
I will check.

adamsilverstein replied:
removed

adamsilverstein committed:
Update packages/editor/src/components/post-author/enhanced-post-author.js

Co-authored-by: Enrique Piqueras <epiqueras@users.noreply.github.com>

adamsilverstein committed:
merge

adamsilverstein committed:
Clean up checks, return a single value from selector.

adamsilverstein committed:
Change element keys to author ids.

adamsilverstein committed:
Refresh the user list when the field value changes or the selection is changed.

adamsilverstein committed:
Refresh the user list when the field value changes or the selection is changed.

adamsilverstein committed:
Remove ‘value’ from author objects.

adamsilverstein committed:
Remove unused id from ComboboxControl.

adamsilverstein committed:
use unshift

adamsilverstein committed:
remove tracking of selectedAuthor

adamsilverstein commented:
@epiqueras Thank you for the feedback. 

I have addressed nearly all of your items and left you some questions. Can you give this a test? In addition to no loading state I noticed we may want to add a "No results found." message as well.

adamsilverstein committed:
merge

adamsilverstein committed:
merge

adamsilverstein committed:
merge

adamsilverstein committed:
Simplify.

adamsilverstein commented:
@epiqueras in the latest commits I changed to extending getAuthors instead of adding a new searchAuthors. 

One issue is that because the `authors`  key in the store is updated, it also updates the entire state of anything depending on authors - creating unintended side effects.  For example in `PostAuthorCheck` if there are fewer than 2 authors, the author selector is not shown - so if you type a search that fails, authors becomes empty, and the field disappears. I removed the check temporarily. 

Another example - I'd like to show a regular select for a small number of authors in the system (eg <=50), with that logic in place searching would again limit the authors and the searchable ui would change to a dropdown.

To fix these issues temporarily  I introduced the query string to the store key, which feels fragile.

Can you please review and see if there is a better way to use `getAuthors` or if you agree we need a separate searching selector/resolver?

Also can you suggest the best approach for tracking the `isLoadingSearchResilts` state we need to show an overlay? Assuming I use the store approach for resolving searches, I think I can yield load loading state from the resolver before and after it makes the data call.

epiqueras commented:
I replied.

adamsilverstein commented:
@epiqueras Thanks for the feedback, I will work on refining and let you know when I have this updated. 

Since the overall approach seems solid, I am going to work on the page selector as well. In that case the selection will be hierarchical and that needs to be displayed in the combo box. Feels like the markup is significantly different so it may make sense to add a hierarchical combo box component.

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein committed:
Use state.users.byId in getAuthor.

adamsilverstein committed:
Track isLoading and latest available authors: grey out existing list during load.

adamsilverstein committed:
Use stableKey from getQueryParts.

adamsilverstein committed:
Add isLoading with useSelect..

adamsilverstein committed:
Add they grayed out style.

adamsilverstein committed:
add an is-loading class when cobocox data is loading.

adamsilverstein committed:
Add an is-loading class when loading

adamsilverstein committed:
Improve component import.

adamsilverstein commented:
@epiqueras This is ready for another review when you have a chance. Feel free to push updates directly to the branch.

epiqueras commented on packages/components/src/combobox-control/index.js:
```diff
@@ -30,15 +33,19 @@ export default function ComboboxControl( {
 		selectedItem,
 	} = useCombobox( {
-		initialSelectedItem: items[ 0 ],
+		initialSelectedItem: items[ initialHighlightedIndex || 0 ],
```
This contradicts the Downshift API. If this is the desired behavior, then you need an `initialSelectedItem` prop, not `initialHighlightedIndex`.

adamsilverstein replied:
I will try using `initialSelectedItem` directly again.

adamsilverstein replied:
improved with `initialSelectedItem` in https://github.com/WordPress/gutenberg/pull/23237/commits/d700436a80faba003c4cbb7b60f3f64b5d19f88. Also fixed the comparison so objects work and the current selected author is now correctly marked with a checkmark in the dropdown.

epiqueras commented on packages/editor/src/components/post-author/check.js:
```diff
@@ -16,10 +16,9 @@ import PostTypeSupportCheck from '../post-type-support-check';
-	authors,
 	children,
 } ) {
-	if ( ! hasAssignAuthorAction || authors.length < 2 ) {
```
What's the rationale for this?

adamsilverstein replied:
hmm, possibly removed by mistake, will check

adamsilverstein replied:
This was related to having the authors reset by search which is now fixed with the query key.  Reverted these changes and verified search with empty search doesn't make the field disappear as before.

epiqueras commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,71 +1,137 @@
-export class PostAuthor extends Component {
-	constructor() {
-		super( ...arguments );
+let latestAvailableAuthors;
```
This won't work when the component is used more than once or when a concurrent mode is enabled.

adamsilverstein replied:
Good point. This was added as a way to capture the available authors  so we can display them greyed out during a new data load (after typing more into the search). I'll remove this for now and we can work to address the UI issues separately.

adamsilverstein replied:
removed this, added an initial spinner instead to display while loading new results.

epiqueras commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,71 +1,137 @@
-		super( ...arguments );
+let latestAvailableAuthors;
+function PostAuthor() {
+	const authors = useSelect( ( select ) => select( 'core' ).getAuthors() );
+	const postAuthor = useSelect( ( select ) =>
+		select( 'core' ).getAuthor(
+			select( 'core/editor' ).getEditedPostAttribute( 'author' )
+		)
```
There is no need for two `useSelect` calls here.

adamsilverstein replied:
really, i thought this was the pattern I saw elsewhere in GB. we need two separate values - the authors and the post author. 

so instead I should to a single useSelect that selects and returns both them restructure them?

adamsilverstein replied:
@gwwar That makes sense, thank you for the code sample which helped me understand the request. Used slightly modified to resolve in [`66f023c` (#23237)](https://github.com/WordPress/gutenberg/pull/23237/commits/66f023c2daba60aebd793e6de8f8eb5874f4e0cf).

epiqueras commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,71 +1,137 @@
+			return {
+				key: author.id,
+				name: author.name,
+				id: author.id,
```
You don't need the `id` property if the `key` has the same value.

adamsilverstein replied:
👍 will remove. we need the key.

adamsilverstein replied:
Removed these.

epiqueras commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,71 +1,137 @@
+	);
+
+	// The currently field value.
+	const [ fieldValue, setFieldValue ] = useState( postAuthor?.name );
```
If you are keeping the value in the local state, why isn't the component controlled?

adamsilverstein replied:
The component already controls its own state and I had trouble trying to make it a controlled component when setting a value, the error was about changing from a controlled to non controlled component. passing a value doesn't seem to work as expected with the controlled combobox, maybe something to address in a follow up? 

The main reason to track state here is to trigger a search when fieldValue changes.

epiqueras replied:
Have you tried passing `fieldValue` to the Combobox?

adamsilverstein replied:
I'm not sure how this would help? Can you explain what you mean here?

adamsilverstein commented on packages/core-data/src/resolvers.js:
```diff
@@ -25,15 +25,34 @@ import {
+	} );
+	const { stableKey } = getQueryParts( query );
+
+	yield receiveUserQuery( `authors::${ stableKey }`, users );
```
so just use `${ stableKey }` - I think I tried this, I'll give it another try.

adamsilverstein replied:
👍 worked

adamsilverstein commented on packages/core-data/src/resolvers.js:
```diff
@@ -25,15 +25,34 @@ import {
-export function* getAuthors() {
+export function* getAuthors( query = {} ) {
+	const path = addQueryArgs( '/wp/v2/users/?who=authors&per_page=100', {
+		...query,
```
👍 will fix

adamsilverstein replied:
removing this spread breaks searching - as we pass `{ search: [string] }` to `getAuthors`. we could remove if I set this here to `'search': query`, is that preferable?  this would allow only searching and not passing other query parameters.

adamsilverstein replied:
@gwwar Thanks for the feedback, I'll give that a try. I think the main issue is making sure the query arrives correctly because the API endpoint expects `{ search: "string" }`.

adamsilverstein replied:
Fixed in [`d97d175` (#23237)](https://github.com/WordPress/gutenberg/pull/23237/commits/d97d175d75b4ebfa5be6131d3023b5576e837ce1)

adamsilverstein commented on packages/core-data/src/selectors.js:
```diff
@@ -75,6 +77,18 @@ export const getUserQueryResults = createSelector(
+ * Returns an Author by ID.
+ *
+ * @param {Object} state   Data state.
+ * @param {string} queryID Query ID.
```
👍 will update

adamsilverstein replied:
updated naming here.

noahtallen committed:
Editor still loads if page or template type is not set

noahtallen committed:
Create homepage state reducer

noahtallen committed:
Remove initial state from PHP

noahtallen committed:
Move home page resolution to client

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein committed:
Keys can be empty.

adamsilverstein committed:
Remove extraneous spread

adamsilverstein committed:
Update naming to authorID, restore spread.

adamsilverstein committed:
Switch to `initialSelectedItem`; improve selected item matching with `isEqual`.

adamsilverstein committed:
Revert changes to post-author check.

adamsilverstein committed:
Remove latestAvailableAuthors and use spinner instead for load.

adamsilverstein commented:
@epiqueras This is ready for another review, thanks again for all the feedback.

noahtallen committed:
Consolidate entityId logic

noahtallen committed:
Consolidate activePage?.path

noahtallen committed:
Rework setupstate

noahtallen committed:
Load home data over API

adamsilverstein committed:
cleanup

noahtallen committed:
Rename camelcase

noahtallen committed:
Remove templateId from setPage

noahtallen committed:
Naive cache results of findTemplate

noahtallen committed:
Rename cache variable

noahtallen committed:
Add text domain to setting translation

noahtallen committed:
Fix existing JS unit tests

noahtallen committed:
Fix e2e test

gwwar reviewed (commented):
I left a few minor code notes. @epiqueras was main blocking work here avoiding component state in post-author?

adamsilverstein committed:
merge from origin/master

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein committed:
update after merge  from origin/master

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein committed:
Revert whitespace change.

adamsilverstein committed:
Remove unnecessary spread.

adamsilverstein committed:
Simplify to single useSelect.

adamsilverstein commented on packages/editor/src/components/post-author/index.js:
```diff
@@ -1,71 +1,132 @@
-	} ) ),
-	withInstanceId,
-] )( PostAuthor );
+export default PostAuthor;
```
@gwwar - I can't recall, I think I changes this as part of making it a functional component. If it is preferred, I can switch this back to using the compose approach.

gwwar replied:
This was an honest question here :) I wasn't sure if we switched patterns again in the project.

adamsilverstein commented on packages/editor/src/store/actions.js:
```diff
@@ -722,7 +722,7 @@ export function updateEditorSettings( settings ) {
 
 const getBlockEditorAction = ( name ) =>
-	function* ( ...args ) {
+	function*( ...args ) {
```
Fixed in [`a7a79b5` (#23237)](https://github.com/WordPress/gutenberg/pull/23237/commits/a7a79b57c40cb5509beafb57f123ab56316ec704)

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein committed:
Fixes for js lint.

adamsilverstein committed:
Ensure the dropdown list is outlined when non-empty.

adamsilverstein commented:
@epiqueras & @gwwar - thank you for your continued feedback, I have updated the PR to address issues you raised. I addressed the missing outline for the dropdown and also have one follow up question about the [use of compose](https://github.com/WordPress/gutenberg/pull/23237#discussion_r470720300) and whether to reintroduce that.

Appreciate your review again when you have a chance.

adamsilverstein committed:
Restore use of compose.

adamsilverstein commented:
Switched back to compose. I tried to get the tests working and could use some help please. I believe we need to mock the stores to get the tests to load the component. 'core/data' is available but 'core is not.

I tried adding something like this in the test:

```js
registerGenericStore( 'core', {
	getActions: () => ( { foo: 'bar' } ),
	subscribe: () => ( { foo: 'bar' } ),
	getSelectors: () => {
		return { 'getAuthors': () => authors };
	},
} );
```

and only got so far before getting blocked with unclear error messages.

adamsilverstein commented:
Pass at tests: https://gist.github.com/adamsilverstein/4ca24e203b1d3143d0a9fcbd50afb0d3 - currently getting an error trying to render the component:
`Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format... in WithSelect(WithInstanceId(PostAuthorCheck))"`

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

gwwar reviewed (commented):
Leaving a higher level review for folks. I think we may need a bit of more polish, but I'm really excited by the potential performance improvement here. cc @epiqueras @ellatrix @youknowriad  or whomever has cycles on the core editor team for a +1 review.

### Context: this will improve performance on large sites

First off, thanks @adamsilverstein for continuing work here! 💖 Getting a search box in for authors will have a great impact on performance and overall usability. 

For context on why we need this PR, the current author dropdown requests ***every single author in batches of 100*** on a site. This is a problem when we have a lot of authors. The component may take **minutes** to load, sometimes causing users to think that either the component is **broken or removed**.

| Before requests completed  | All requests completed | So many author requests 😱 |
| ------------- | ------------- |------------- |
| <img width="279" alt="where is author control?" src="https://user-images.githubusercontent.com/1270189/91254146-62bbe800-e716-11ea-81cd-8f9c7bab912a.png"> | <img width="289" alt="after all requests complete" src="https://user-images.githubusercontent.com/1270189/91254155-67809c00-e716-11ea-93f9-aa952d36073b.png"> | <img width="1426" alt="Screen Shot 2020-08-25 at 9 01 33 PM" src="https://user-images.githubusercontent.com/1270189/91254179-7c5d2f80-e716-11ea-957e-0e29d2b71c83.png"> |

For curious folks this happens in code when we request in apiFetch a `per_page` value of `-1`
```
path: `/wp/v2/users/${ id }?who=authors&per_page=1`,
```

There's a middleware that handles the follow up requests.
https://github.com/WordPress/gutenberg/blob/master/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L39-L45

With this PR behavior is now updated to request the first 100 authors on load, even if we have thousands of authors. This makes it so the author component **is available for interaction on a much more consistent basis**.

| < 100 | > 100 | Dropdown showing default 100 results |
| ------------- | ------------- |------------- |
| <img width="1416" alt="existing behavior   100" src="https://user-images.githubusercontent.com/1270189/91254345-03aaa300-e717-11ea-945c-d772b53f3cdc.png"> | <img width="1416" alt="only load first 100" src="https://user-images.githubusercontent.com/1270189/91254351-09a08400-e717-11ea-9e01-c8500ac41365.png"> | <img width="304" alt="Screen Shot 2020-08-25 at 8 48 14 PM" src="https://user-images.githubusercontent.com/1270189/91254357-0e653800-e717-11ea-8b99-45217df02599.png"> |

### Minor Feedback on Interactions

We could use a bit of polish or feedback on the interaction with the author dropdown while searching. Maybe we can drop the PR into the design channel for a review?

![search-behavior](https://user-images.githubusercontent.com/1270189/91254586-9cd9b980-e717-11ea-899b-2358975390f0.gif)

Some things that stood out to me:
* When I select this (not directly clicking the chevron), I'm unsure of what to do next. Should we show a dropdown of results too, or select the search text?
* Search results feel sluggish even on a local dev env. We can play with some perceived perf options here or tune the debounce.
* Typing in the exact author name, requires me to still select the dropdown option before I can save the post.
* I'm sure folks may also have some styling suggestions.

Since the performance impact is very large for a good enough implementation, I'm more inclined to note that I'd recommend some of this feedback as follow up PRs targeted to the same release, if it'd help with movement on this PR.

### How do we E2E test both author states?

Most E2E tests are simple and rely on actions on a mostly blank site to build up content and testing.

If we want to test a site with <100 authors, and another with >100 authors, do we currently have tooling to allow this, or do we need to add this?

### Overall

I did some quick smoke testing and all the basics are here. We do need to polish the interactions and maybe expand on testing in follow up PRs, but I'd highly recommend that try to get an implementation in sooner rather than later, as the current experience on large sites is detrimental enough that people may need to switch to the classic editor for a basic function like updating a post author.

youknowriad commented on packages/core-data/src/resolvers.js:
```diff
@@ -25,15 +25,37 @@ import {
+ */
+export function* getAuthors( query = {} ) {
+	const path = addQueryArgs(
+		'/wp/v2/users/?who=authors&per_page=100',
```
Should we remove the per_page argument here? the query can override it if I'm not wrong.

youknowriad replied:
Also it seems that the "who" argument shouldn't be hardcoded like that but merged into "query" in order for the "users per query" state to be populated properly.

It seems like right now, if one does `getUsers` with a similar query object, you might get bad results.

youknowriad replied:
Thinking more potentially we can remove this `getAuthors` selector/resolver as it's just a shortcut for `getUsers` AFAIK.

adamsilverstein replied:
I'll give `getUsers` a try.

youknowriad commented on packages/core-data/src/resolvers.js:
```diff
@@ -25,15 +25,37 @@ import {
+ * @param {string} id The ID of the author to retrieve.
  */
-export function* getAuthors() {
+export function* getAuthor( id = '' ) {
```
Is the default value necessary here?

youknowriad replied:
Actually, do we need this selector/resolver at all, why not just use `getUser` ?

adamsilverstein replied:
Not sure I looked at getUser, I'll try that.

youknowriad commented on packages/core-data/src/resolvers.js:
```diff
@@ -25,15 +25,37 @@ import {
+export function* getAuthor( id = '' ) {
 	const users = yield apiFetch( {
-		path: '/wp/v2/users/?who=authors&per_page=-1',
+		path: `/wp/v2/users/${ id }?who=authors&per_page=1`,
```
Is the per_page argument necessary since we're explicitly providing an id?

adamsilverstein replied:
probably not, more habit of limiting queries when possible. I can remove.

adamsilverstein commented:
@gwwar Thanks for the high level overview and continued attention to this ticket. Completely agree of the UX/UI points and room for improvement. The upstream maintainer of Downshift also seems very open to improvements if we find issues.

> Getting a search box in for authors will have a great impact on performance and overall usability.

I'd like to mention that the page selector for pages suffers from the exact same issue when sites have many pages (https://github.com/WordPress/gutenberg/issues/9441) - I already have a branch in progress fixing that selector using the Downshift. Once we nail down the technical details here, I'll open a separate PR for that change.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

youknowriad committed:
Merge remote-tracking branch 'origin/master' into add/7385-author-combobox-2

youknowriad committed:
Avoid getAuthor and getAuthors

youknowriad committed:
deprecate getAuthors

youknowriad committed:
Small refactoring

youknowriad commented:
I made some small changes here:

 - Removed the getAuthor/getAuthors and just used getUser/getUsers
 - Refactored the PostAuthor component logic a bit (I hope I managed to simplify/clarify it a bit)

I think there's still some work design wise. It seems to use the `CustomSelectControl` component but the style of the component look very different and not consistent with other controls (SelectControl, ...), so we may need some design eyes to make things more consistent cc @jasmussen @ItsJonQ 

Also, the label is displayed on its own line by default for all controls, maybe we need a prop to support inline labels as it seems that it's what we want here.

adamsilverstein commented:
> I made some small changes here

Thank you for adjusting those!

> Also, the label is displayed on its own line by default for all controls, maybe we need a prop to support inline labels as it seems that it's what we want here.

Great idea.

adamsilverstein commented:
@youknowriad thank you for the updates, I'll give the latest version a test.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

jasmussen committed:
Polish.

jasmussen commented:
I pushed some polish in https://github.com/WordPress/gutenberg/pull/23237/commits/2383f1cf583b7c8028fea7891681d0e1f6b239ef. 

![polish](https://user-images.githubusercontent.com/1204802/92583007-028a8180-f292-11ea-9dd4-e5167b955e3b.gif)

There's a lot going on here, and there should be less. But this at least moves us a fair bit closer to what it should look like, and it fixes the off colors, radii, paddings and offsets from before:

![dropdown](https://user-images.githubusercontent.com/1204802/92583130-2ea60280-f292-11ea-9b03-22fdf479dc2e.gif)

I'd love to see us have a horizontal layout more like this:

<img width="261" alt="Screenshot 2020-09-09 at 11 47 08" src="https://user-images.githubusercontent.com/1204802/92583160-39609780-f292-11ea-8f09-c84381582544.png">

But I have to move on to some other PRs at the moment.

jasmussen commented:
Forgot to clarify: really nice work here! And while I'd love to see more design love applied to the new select box, this is probably a good start so the PR doesn't drag on forever!

adamsilverstein commented:
The visual improvements here are great, thank you @jasmussen!

Noting that I received an error when testing this:

`Uncaught TypeError: Cannot read property 'name' of undefined`

To reproduce this:
* Ensure you have over 100 users.
* Select a user beyond user 100 and update the post.
* Refresh the page.

The source of the error I believe is that the selected post author is not being correctly fetched for the post (since it is not among the initial 100 fetched, we need to explicitly fetch it). I'll try to dig into this later unless @youknowriad beats me to it.

adamsilverstein committed:
Only set field value once postAuthor is available.

adamsilverstein committed:
Onle set field value once postAuthor is available.

adamsilverstein commented:
Fixed the error noted in my previous comment in [`9428906` (#23237)](https://github.com/WordPress/gutenberg/pull/23237/commits/9428906e7776446483d8e229887d904d921ee930). When the post author is not in the initial set of authors, postAuthor is not defined initially. Waiting until it is defined to set the field value fixes the issue.

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' of github.com:WordPress/gutenberg

adamsilverstein committed:
Merge branch 'master' into add/7385-author-combobox-2

adamsilverstein commented:
Started similar approach to parent page selector which has the same issue with large numbers of pages: https://github.com/WordPress/gutenberg/pull/25267

youknowriad committed:
Remove useless test

youknowriad commented:
Noticed a weird glitch here (I'm testing on safari), When I click on one of the select options (and before releasing the button), the panel chevron icons on the sidebar move to the right.

youknowriad committed:
Merge branch 'add/7385-author-combobox-2' of github.com:WordPress/gutenberg into add/7385-author-combobox-2

adamsilverstein commented:
> Noticed a weird glitch here (I'm testing on safari), When I click on one of the select options (and before releasing the button), the panel chevron icons on the sidebar move to the right.

Does it also affect the font size dropdown selector where the combobox is already used?

youknowriad commented:
No, it only affects the author control, that said the font size one is not a ComboboxControl, it's a CustomSelectControl.

adamsilverstein commented:
> No, it only affects the author control, that said the font size one is not a ComboboxControl, it's a CustomSelectControl.

Oh, I see now. I thought previously that the selector used `ComboboxControl`, maybe that changed?

afercia commented:
For a first accessibility feedback, please see this comment on the related PR for the Page parent combobox: https://github.com/WordPress/gutenberg/pull/25267#issuecomment-692757270

adamsilverstein commented:
@afercia I took a pass at starting implementing the author selector using the same component the tag autocomplete selector uses: `FormTokenField` in https://github.com/WordPress/gutenberg/pull/25420

It sort of works and illustrates some of the fundamental UX differences here:

* The tag selector never shows a dropdown of selections, Downshift shows a default list of items (maybe this is ok - for <= 50 users we can show a standard select element)
* The tag selector is designed to contain multiple tags, the page/author selectors are designed to choose a single item. I hack that by hiding the existing tag, with mixed results.
* The tag selector is designed to allow you to "clear" all of the tags and have a blank field. This should be possible for pages (no parent) but **not** authors.
* one other limitation - values in the selector are flat, not objects, so we can't store both the author name and id. instead we combine the two and extract them using a regex when selected.

Given the fact that this component already has excellent accessibility, does it seem worth trying to extend the underlying component to better fit our use case?

afercia commented:
@adamsilverstein thanks for further exploring this issue! 

Well if you ask me, I see a value in abstracting the Gutenberg combobox (auto-completer) implementation and make it more reusable, not only for internal usage but also for usage by plugins. I think I mentioned this a few times in other issues 🙂 As per the technical side, I'm afraid I don't have the skills to be of help with coding.

> The tag selector never shows a dropdown of selections, Downshift shows a default list of items (maybe this is ok - for <= 50 users we can show a standard select element)

Using a native `<select>` element for a small amount of users seems a very good idea. I'm not sure a search / filtering / suggestions UI would make sense for, say, 5 users even from a visual perspective 🙂

youknowriad committed:
Merge remote-tracking branch 'origin/master' into add/7385-author-combobox-2

youknowriad committed:
Use the new combobox control

youknowriad commented:
I updated this to use the new ComboboxControl. 
I had to update the component a bit cause its behavior didn't match the real use-case expectations but I believe it's good now.

youknowriad commented:
I think we can explore isLoading prop on a separate PR to get it right.

adamsilverstein commented:
🎉 

Thank you @youknowriad for working on this. I'll give this final version a test to see if I notice any issues. I have been working on these selectors for a long time (years?!) so it is really exciting to see this get fixed!

adamsilverstein commented:
Confirming this works well in my test, excited to see this land!

adamsilverstein commented:
> @adamsilverstein thanks for further exploring this issue!
> 
> Well if you ask me, I see a value in abstracting the Gutenberg combobox (auto-completer) implementation and make it more reusable, not only for internal usage but also for usage by plugins. I think I mentioned this a few times in other issues 🙂 As per the technical side, I'm afraid I don't have the skills to be of help with coding.
> 
> > The tag selector never shows a dropdown of selections, Downshift shows a default list of items (maybe this is ok - for <= 50 users we can show a standard select element)
> 
> Using a native `<select>` element for a small amount of users seems a very good idea. I'm not sure a search / filtering / suggestions UI would make sense for, say, 5 users even from a visual perspective 🙂

@afercia We addressed this in https://github.com/WordPress/gutenberg/pull/26426 for the post author selector (and I plan to do the same for the page parent selector soon). Thanks again for all your guidance on this issue.
